aip/client-libraries/4222.md

```diff
@@ -13,26 +13,172 @@ routing.
 
 ## Guidance
 
-Code generators **should** look at URI-based variables declared in the
+Code generators **must** use the annotations to generate client libraries that,
+on a per-RPC basis, extract routing information from the request payload and
+add that information to the routing header.
+
+There are two annotations that specify how to extract routing information from
+the request payload:
+
+  * the [`google.api.routing`][routing] annotation that specifies how to construct 
+routing headers explicitly 
+  * the [`google.api.http`][http] annotation that may specify how to construct 
+routing headers implicitly.
+
+For any given RPC, if the explicit routing headers annotation is present, the code
+generators **must** use it and ignore any routing headers that might be implicitly
+specified in the [`google.api.http`][http] annotation. If the explicit routing
+headers annotation is absent, the code generators **must** parse the
+[`google.api.http`][http] annotation to see if it specifies routing headers
+implicitly, and use that specification.
+
+## Explicit Routing Headers (`google.api.routing`)
+
+For an unary or server-streaming RPC the code generator **must** look at the routing
+parameters specified in the [`google.api.routing`][routing] annotation, if present.
+Any given routing parameter specifies a field name and a pattern with exactly one
+named resource ID path segment. For example:
+
+```proto
+rpc CreateTopic(CreateTopicRequest) {
+  option (google.api.routing) = {
+    routing_parameters {
+      field: "parent"
+      path_template: "{project=projects/*}/**"
+    }
+  }
+}
+```
+**Note:** An empty `google.api.routing` annotation is acceptable. It means that no
+routing headers should be generated for the RPC, when they otherwise would be
+e.g. implicitly from the `google.api.http` annotation.
+
+**Note:** It is acceptable to omit the pattern in the resource ID segment, `{parent}`
+for example, is equivalent to `{parent=*}` and **must** be parsed, e.g.:
+```proto
+routing_parameters {
+  field: "parent"
+  path_template: "projects/{parent}"
+}
+```
+
+is the same as
+
+```proto
+routing_parameters {
+  field: "parent"
+  path_template: "projects/{parent=*}"
+}
+```
+
+**Note** It is acceptable to omit the `path_template` field altogether. An omitted 
+`path_template` is equivalent to a `path_template` with the same resource ID name as
+the field and the pattern `**`, and **must** be parsed, e.g.:
```

Comment:
Possibly explicitly note that this excludes empty strings? It comes up later, but I certainly wondered about it when reading this part.

Reply:
:+1: